File: tests/cpp-tests/Classes/AccelerometerTest/AccelerometerTest.cpp
#include "AccelerometerTest.h"
#include "../testResource.h"
#include "cocos2d.h"

#define FIX_POS(_pos, _min, _max) \
    if (_pos < _min)        \
    _pos = _min;        \
else if (_pos > _max)   \
    _pos = _max;        \

//------------------------------------------------------------------
//
// AccelerometerTest
//
//------------------------------------------------------------------
AccelerometerTest::AccelerometerTest()
{
}

AccelerometerTest::~AccelerometerTest()
{
    _ball->release();
    Device::setAccelerometerEnabled(false);
}

std::string AccelerometerTest::title() const
{
    return "AccelerometerTest";
}

void AccelerometerTest::onEnter()
{
    Layer::onEnter();

    Device::setAccelerometerEnabled(true);
    auto listener = EventListenerAcceleration::create(CC_CALLBACK_2(AccelerometerTest::onAcceleration, this));
    _eventDispatcher->addEventListenerWithSceneGraphPriority(listener, this);

    auto label = Label::createWithTTF(title().c_str(), "fonts/arial.ttf", 32.0f);
    addChild(label, 1);
    label->setPosition( Vec2(VisibleRect::center().x, VisibleRect::top().y-50) );

    _ball = Sprite::create("Images/ball.png");
    _ball->setPosition(Vec2(VisibleRect::center().x, VisibleRect::center().y));
    addChild(_ball);

    _ball->retain();
}

void AccelerometerTest::onAcceleration(Acceleration* acc, Event* event)
{
    /*FIXME: Testing on the Nexus S sometimes _ball is NULL */
    if ( _ball == nullptr ) {
        return;
    }

    auto ballSize  = _ball->getContentSize();

    auto pos  = _ball->getPosition();

    pos.x += acc->x * 9.81f;
    pos.y += acc->y * 9.81f;

    FIX_POS(pos.x, (VisibleRect::left().x+ballSize.width / 2.0), (VisibleRect::right().x - ballSize.width / 2.0));
    FIX_POS(pos.y, (VisibleRect::bottom().y+ballSize.height / 2.0), (VisibleRect::top().y - ballSize.height / 2.0));
    _ball->setPosition(pos);
}

//------------------------------------------------------------------
//
// AccelerometerTestScene
//
//------------------------------------------------------------------
void AccelerometerTestScene::runThisTest()
{
    auto layer = new AccelerometerTest();
    addChild(layer);
    layer->release();

    Director::getInstance()->replaceScene(this);
}

File: tests/lua-tests/project/proj.tizen/res/src/AccelerometerTest/AccelerometerTest.lua
local function AccelerometerMainLayer()

    local function title()
      return "AccelerometerTest"
    end
    local layer = cc.Layer:create()

    local function onEnter()
        layer:setAccelerometerEnabled(true)
        local label = cc.Label:createWithTTF(title(), "fonts/arial.ttf", 32)
        layer:addChild(label, 1)
        label:setAnchorPoint(cc.p(0.5, 0.5))
        label:setPosition( cc.p(VisibleRect:center().x, VisibleRect:top().y - 50) )

        local ball = cc.Sprite:create("Images/ball.png")
        ball:setPosition(cc.p(VisibleRect:center().x, VisibleRect:center().y))
        layer:addChild(ball)

        local function accelerometerListener(event,x,y,z,timestamp)
            local target  = event:getCurrentTarget()
            local ballSize = target:getContentSize()
            local ptNowX,ptNowY    = target:getPosition()
            ptNowX = ptNowX + x * 9.81
            ptNowY = ptNowY + y * 9.81

            local minX  = math.floor(VisibleRect:left().x + ballSize.width / 2.0)
            local maxX  = math.floor(VisibleRect:right().x - ballSize.width / 2.0)
            if ptNowX <   minX then
                ptNowX = minX
            elseif ptNowX > maxX then
                ptNowX = maxX
            end
      
            local minY  = math.floor(VisibleRect:bottom().y + ballSize.height / 2.0)
            local maxY  = math.floor(VisibleRect:top().y   - ballSize.height / 2.0)
            if ptNowY <   minY then
                ptNowY = minY
            elseif ptNowY > maxY then
                ptNowY = maxY
            end

            target:setPosition(cc.p(ptNowX , ptNowY))
        end

        local listerner  = cc.EventListenerAcceleration:create(accelerometerListener)
        layer:getEventDispatcher():addEventListenerWithSceneGraphPriority(listerner,ball)
    end

    local function onExit()
        layer:setAccelerometerEnabled(false)
    end

    local function onNodeEvent(event)
        if "enter" == event then
            onEnter()
        elseif "exit" == event then
            onExit()
        end
    end

    layer:registerScriptHandler(onNodeEvent)
    
    return layer
end


function AccelerometerMain()
	cclog("AccelerometerMain")
	local scene = cc.Scene:create()
	scene:addChild(AccelerometerMainLayer())
	scene:addChild(CreateBackMenuItem())
	return scene
end
